### DISCUSSION TITLE
Sha collision

### DISCUSSION THREAD

**Discussion by vbjay** [git-lfs/git-lfs#5322] [Q&A]
With git lfs storage being shared if configured, does git lfs handle sha collisions at all?  In one repo I can see where it might not be a problem but the potential of collisions grows when many repos use a shared git lfs storage.

**Comment by chrisd8088** [comment]
On the client side, the SHA-256 OIDs generated by Git LFS are specific to a given repository.

So the question here is really about how Git LFS storage providers design their services, and in general, they should keep each repository's set of Git LFS objects separate from others (although in practice the chance of a collision is very low).

Taking GitHub as an example, the URLs for their storage of Git LFS objects look like the following, where the `<id>` is specific to a given repository, so there is no chance of collision between objects across repositories.
```
https://github-cloud.s3.amazonaws.com/<prefix>/<id>/a6/e2/a6e200708d54745f00ec04050da9440fdbb0a77651de113003418889e301fef6
```

**Reply by vbjay** [reply]
Thanks. Was definitely more concerned with the 2 cases of 

- client and user says storage = some shared folder path that mingles lfs objects
- server uses storage to do the same thing where the lfs objects are mingled with multiple repos

Think of the case where file x and file y happen to have a collision but different content. Yes I know it is unlikely but still a chance.  Yes I know you have the chance of files in x repos of different content actually hashing to the same hash.

What I'm wondering is if git lfs has any mitigation or do you just get a random blob back between the 2 or last modified version or what?

**Reply by chrisd8088** [reply]
The SHA-256 hash is considered to be unique within a repo; there is no special mitigation for a collision.  You are much more likely to see a collision with Git's SHA-1 hash than with Git LFS's.  SHA-256 collisions are, for all intents and purposes, not something to worry about right now; there are a few resources out there which explain how incredibly unlikely a collision is.
